arkane/statmech.py

```diff
@@ -969,128 +969,30 @@ def determine_qm_software(fullpath):
-                        # avoid dividing by 0, but check we do extrapolate to 0 within the tolerance
-                        if abs(y0) < epsilon:
-                            return True
-                        else:
-                            return False
-                elif y[0] != y[1]:
-                    x0 = x[0] + (y[i] - y[0]) * (x[1] - x[0]) / (y[1] - y[0])
-                    if x[i] != 0:
-                        if nearly_equal(a=x0, b=x[i]):
-                            return True
-                        else:
-                            return False
-                    else:
-                        if abs(x0) < epsilon:
-                            return True
-                        else:
-                            return False
-                else:
-                    raise StatmechError('Could not determine whether species {0} is linear, something must be wrong '
-                                        'with its coordinates:\n{1}'.format(label, coordinates))
-    else:
-        # this problem cannot be reduced, and must have a full 3D treatment
-        vector_list = []
-        for i, _ in enumerate(x):
-            # generate a vector list where each element in the list contains [v1, v2]
-            # which are vectors (subtracting two points in space) between atoms A-B and B-C (or i, i-1 and i-1, i-2)
-            if i > 1:  # only start the algorithm at the third atom (index 2)
-                v1 = [x[i] - x[i - 1], y[i] - y[i - 1], z[i] - z[i - 1]]  # between points A, B
-                v2 = [x[i - 1] - x[i - 2], y[i - 1] - y[i - 2], z[i - 1] - z[i - 2]]  # between points B, C
-                vector_list.append([v1, v2])
-        for v1, v2 in vector_list:
-            if v1[0] and v2[0]:
-                ax = v1[0] / v2[0]  # This is how much the x coord of point C is stretched relative to the AB vector
-            else:
-                ax = 0
-            if v1[1] and v2[1]:
-                ay = v1[1] / v2[1]
-            else:
-                ay = 0
-            if v1[2] and v2[2]:
-                az = v1[2] / v2[2]
-            else:
-                az = 0
-            if ax and ay and not nearly_equal(ax, ay):
-                break
-            if ax and az and not nearly_equal(ax, az):
-                break
-            if ay and az and not nearly_equal(ay, az):
-                break
-        else:
-            return True
-        return False
+    # A tensor containing all distance vectors in the molecule
+    d = -numpy.array([c[:, numpy.newaxis] - c[numpy.newaxis, :] for c in coordinates.T])
+    for i, c in enumerate(coordinates):
+        if i > 1:
+            u1 = d[:, 0, 1] / numpy.linalg.norm(d[:, 0, 1])  # unit vector between atoms 0 and 1
+            u2 = d[:, 1, i] / numpy.linalg.norm(d[:, 1, i])  # unit vector between atoms 1 and i
+            a = math.degrees(numpy.arccos(numpy.clip(numpy.dot(u1, u2), -1.0, 1.0)))  # angle between atoms 0, 1, i
```

Comment:
Are you using `numpy.clip` to avoid invalid arguments for `numpy.arccos` if there are floating point errors in the calculation of the dot product?

Reply:
Yes, I thought to use cautious although my examples ran fine without it